src/databricks/sqlalchemy/dialect/__init__.py

```diff
@@ -0,0 +1,300 @@
+"""This module's layout loosely follows example of SQLAlchemy's postgres dialect
+"""
+
+import decimal, re, datetime
+from dateutil.parser import parse
+
+from sqlalchemy import types, processors
+from sqlalchemy.engine import default
+from sqlalchemy.exc import DatabaseError
+from sqlalchemy.engine import reflection
+
+from databricks import sql
+
+
+from databricks.sqlalchemy.dialect.base import (
+    DatabricksDDLCompiler,
+    DatabricksIdentifierPreparer,
+)
+from databricks.sqlalchemy.dialect.compiler import DatabricksTypeCompiler
+
+try:
+    import alembic
+except ImportError:
+    pass
+else:
+    from alembic.ddl import DefaultImpl
+
+    class DatabricksImpl(DefaultImpl):
+        __dialect__ = "databricks"
+
+
+class DatabricksDecimal(types.TypeDecorator):
+    """Translates strings to decimals"""
+
+    impl = types.DECIMAL
+
+    def process_result_value(self, value, dialect):
+        if value is not None:
+            return decimal.Decimal(value)
+        else:
+            return None
+
+
+class DatabricksTimestamp(types.TypeDecorator):
+    """Translates timestamp strings to datetime objects"""
+
+    impl = types.TIMESTAMP
+
+    def process_result_value(self, value, dialect):
+        return processors.str_to_datetime(value)
+
+    def adapt(self, impltype, **kwargs):
+        return self.impl
+
+
+class DatabricksDate(types.TypeDecorator):
+    """Translates date strings to date objects"""
+
+    impl = types.DATE
+
+    def process_result_value(self, value, dialect):
+        return processors.str_to_date(value)
+
+    def adapt(self, impltype, **kwargs):
+        return self.impl
+
+
+class DatabricksDialect(default.DefaultDialect):
+    """This dialect implements only those methods required to pass our e2e tests"""
+
+    # Possible attributes are defined here: https://docs.sqlalchemy.org/en/14/core/internals.html#sqlalchemy.engine.Dialect
+    name: str = "databricks"
+    driver: str = "thrift"
+    default_schema_name: str = "default"
+
+    preparer = DatabricksIdentifierPreparer
+    type_compiler = DatabricksTypeCompiler
+    ddl_compiler = DatabricksDDLCompiler
+    supports_statement_cache: bool = True
+    supports_multivalues_insert: bool = True
+    supports_native_decimal: bool = True
+    supports_sane_rowcount: bool = False
+
+    @classmethod
+    def dbapi(cls):
+        return sql
+
+    def create_connect_args(self, url):
+        # TODO: can schema be provided after HOST?
+        # Expected URI format is: databricks+thrift://token:dapi***@***.cloud.databricks.com?http_path=/sql/***
+
+        kwargs = {
+            "server_hostname": url.host,
+            "access_token": url.password,
+            "http_path": url.query.get("http_path"),
+            "catalog": url.query.get("catalog"),
+            "schema": url.query.get("schema"),
+            "_user_agent_entry": "+".join(
+                ["sqlalchemy", url.query.get("user_agent_entry", "")]
+            ),
+        }
```

Comment:
Yes, our partners include their name in the user agent. So for example:

`_user_agent_entry = 'AcmeINC'`

produces this user agent

`PyDatabricksSqlConnector/2.3.x (sqlalchemy + AcmeINC)`

This tells us they connected from Python, with which version, using the sqlalchemy API, and the name of the partner. If we don't give partners a way to provide this, we implicitly stop them from using SQLAlchemy.

Reply:
Sorry, that was not my point. My point was that a PySQL user can pass a bunch more parameters to the driver. Are those parameters not supported if using SQLAlchemy?

Reply:
Great point! I've updated the code so that all underlying connection configurations (including the user agent string) are handled in the way SQLAlchemy suggests [here](https://docs.sqlalchemy.org/en/20/core/engines.html#custom-dbapi-connect-arguments-on-connect-routines). Code is here: 8fa08ea784558726858063608d4b8ae86dfdea11